### ISSUE TITLE
about the calculation of cosine similarity in Figure 2

### ISSUE THREAD

**Issue by LiQiang0307** [dingfengshi/TriDet#13]
Upon reading your paper "TriDet: Temporal Action Detection with Relative Boundary Modeling," I found it to be an outstanding contribution to the field of TAL performance improvement. However, I am uncertain about the calculation of cosine similarity in Figure 2. Could you kindly share the code for this part? Thank you very much, and I eagerly await your response.

![image](https://github.com/dingfengshi/TriDet/assets/44935269/c74740e9-03b0-45fe-bf7d-69f63235133f)

**Comment by dingfengshi** [comment]
Hi! I can share you the code of this part. Can you leave any contact information where the code can be sent to you?

**Comment by LiQiang0307** [comment]
I am delighted to receive your message. Here is my email address:  liq782@nenu.edu.cn
Thank you once more and I wish you great success in your research！☕

**Comment by dingfengshi** [comment]
> I am delighted to receive your message. Here is my email address: [liq782@nenu.edu.cn](mailto:liq782@nenu.edu.cn) Thank you once more and I wish you great success in your research！☕

The code has been sent to you~ Thank you for your wish!

**Comment by QX-N** [comment]
Hello, I also want the code for this part and my email is qxn1248@163.com. Could share it to me?
Best wishes.

**Comment by Damon-Salvatore-liu** [comment]
> > I am delighted to receive your message. Here is my email address: [liq782@nenu.edu.cn](mailto:liq782@nenu.edu.cn) Thank you once more and I wish you great success in your research！☕
> 
> The code has been sent to you~ Thank you for your wish!

Hello, I also want the code for this part and my email is liupengli69@163.com. Could share it to me?
Best wishes.

**Comment by dingfengshi** [comment]
Hi, for anyone who needs it. You can modify the code to implement the visualization of cosine similarity.

### 1. Define a cosine similarity function somewhere.
```python
def viz_cos_feature_distance(x, mask):
    x_valid= x[:, :, mask.squeeze()][0].transpose(0, 1)
    x_mean= torch.mean(x_valid, dim=0).unsqueeze(0)
    cos_sim = torch.cosine_similarity(x_mean, x_valid).mean() 
    return cos_sim
```

### 2. Modify `libs/modeling/meta_archs.py`
**Add two variables in the \_\_init\_\_  of class TriDet**
```python
self.cos_val = 0. 
self._cnt = 0
```
**Add code below after `batched_inputs, batched_masks = self.preprocessing(video_list)` in the `forward` function of TriDet**
```python
self.cos_val += viz_cos_feature_distance(batched_inputs, batched_masks) 
self._cnt += 1
```

### 2. Modify `libs/modeling/backbone.py`
**Initial some variables in the \_\_init\_\_  of class SGPBackbone**
```python
self.cos_val = {}
for i in range(len(self.stem)): 
    self.cos_val['stem{}'.format(i)] = 0.
for i in range(len(self.branch)): 
    self.cos_val['branch{}'.format(i)] = 0.
self._cnt = 0
```

**Modify the `forward` function**

Add `self.cos_val['stem{}'.format(idx)] += viz_cos_feature_distance(x, mask)` after `x, mask = self.stem[idx](x, mask)`

Add `self.cos_val['branch{}'.format(idx)] += viz_cos_feature_distance(x, mask)` after `x, mask = self.branch[idx](x, mask)
`

### 2. Modify `libs/utils/tran_utils.py`
**Modify the `valid_one_epoch` function**

Add below code at the end of the loop `for iter_idx, video_list in enumerate(val_loader, 0):`

```python
print("[input feature]= cos: {}".format(model.module.cos_val / model.module._cnt))
for i in range(len(model.module.backbone.stem)):
    cos_v = model.module.backbone.cos_val['stem{}'.format(i)]
    print("[stem{}] cos: {}".format(i, cos_v / model.module._cnt))

for i in range(len(model.module.backbone.branch)):
    cos_v = model.module.backbone.cos_val['branch{}'.format(i)]
    print("[branch{}]  cos: {}".format(i, cos_v / model.module._cnt))
```
